src/TesApi.Web/TesEventMessage.cs

```diff
@@ -0,0 +1,178 @@
+                throw new ArgumentException("This message has no tags.", nameof(tags));
+            }
+
+            if (tags.ContainsKey(ProcessedTag))
+            {
+                throw new ArgumentException("This message was already processed.", nameof(tags));
+            }
+
+            // There are up to 10 tags allowed. We will be adding one.
+            // https://learn.microsoft.com/azure/storage/blobs/storage-manage-find-blobs?tabs=azure-portal#setting-blob-index-tags
+            if (tags.Count > 9)
+            {
+                throw new ArgumentException("This message does not have space to add the processed tag.", nameof(tags));
+            }
+
+            _azureProxy = azureProxy;
+            _logger = logger;
+            _storageAccessProvider = storageAccessProvider;
+            _uri = blobAbsoluteUri;
+            Tags = tags.AsReadOnly();
+            Event = @event;
+        }
+
+        /// <summary>
+        /// Gets the details of this event message.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<(string Id, AzureBatchTaskState State)> GetMessageBatchStateAsync(CancellationToken cancellationToken)
+        {
+            var messageText = await _azureProxy.DownloadBlobAsync(_uri, cancellationToken);
+            var result = Newtonsoft.Json.JsonConvert.DeserializeObject<Tes.Runner.Events.EventMessage>(messageText);
+
+            // TODO: throw if null
+            // Validate. Suggestions include:
+            //Guid.TryParse(result.Id, out _)
+            //Tes.Runner.Events.EventsPublisher.EventVersion.Equals(result.EventVersion, StringComparison.Ordinal)
+            //Tes.Runner.Events.EventsPublisher.EventDataVersion.Equals(result.EventDataVersion, StringComparison.Ordinal)
+            //Tes.Runner.Events.EventsPublisher.TesTaskRunnerEntityType.Equals(result.EntityType, StringComparison.Ordinal)
+            //Tes.Runner.Events.EventsPublisher.TesTaskRunnerEntityType.Equals(result.EntityType, StringComparison.Ordinal)
+            //Event.Equals(result.Name, StringComparison.Ordinal)
+            //new[] { Tes.Runner.Events.EventsPublisher.StartedStatus, Tes.Runner.Events.EventsPublisher.SuccessStatus, Tes.Runner.Events.EventsPublisher.FailedStatus }.Contains(result.StatusMessage)
+
+            // Event type specific validations
+            //
+
+            _logger.LogDebug("Getting batch task state from event {EventName} for {TesTask}.", result.Name ?? Event, result.EntityId);
+            return (result.EntityId, GetCompletedBatchState(result));
+        }
+
+        /// <summary>
+        /// Marks this event message processed.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task MarkMessageProcessed(CancellationToken cancellationToken)
+        {
+            await _azureProxy.SetBlobTags(
+                _uri,
```

Comment:
What is the lifecycle of this URI?  Does it have the SAS token and if so, is there a risk it could expire?

Reply:
From a single call, all non-processed event blobs are gathered, SAS tokens are obtained, and all events are processed (which consists of updating `TesTask`s in the database), in parallel as much as possible. The only way that the token could expire is essentially if the tes task database were down so long and the db retry calls lasted long enough for the SAS to expire (not likely).